tests/test-exports.c

```diff
@@ -0,0 +1,482 @@
+    {
+      const NotFilesystem *not = &not_filesystems[i];
+      g_autoptr(GError) error = NULL;
+      char *normalized = NULL;
+      FlatpakFilesystemMode mode;
+      gboolean ret;
+
+      g_test_message ("%s", not->input);
+      ret = flatpak_context_parse_filesystem (not->input, &normalized, &mode,
+                                              &error);
+      g_assert_error (error, G_OPTION_ERROR, not->code);
+      g_assert_false (ret);
+      g_assert_null (normalized);
+    }
+}
+
+static void
+test_empty (void)
+{
+  g_autoptr(FlatpakBwrap) bwrap = flatpak_bwrap_new (NULL);
+  g_autoptr(FlatpakExports) exports = flatpak_exports_new ();
+  gsize i;
+
+  g_assert_false (flatpak_exports_path_is_visible (exports, "/run"));
+  g_assert_cmpint (flatpak_exports_path_get_mode (exports, "/tmp"), ==,
+                   FLATPAK_FILESYSTEM_MODE_NONE);
+
+  flatpak_bwrap_add_arg (bwrap, "bwrap");
+  flatpak_exports_append_bwrap_args (exports, bwrap);
+  flatpak_bwrap_finish (bwrap);
+  print_bwrap (bwrap);
+
+  i = 0;
+  g_assert_cmpuint (i, <, bwrap->argv->len);
+  g_assert_cmpstr (bwrap->argv->pdata[i++], ==, "bwrap");
+
+  i = assert_next_is_os_release (bwrap, i);
+
+  g_assert_cmpuint (i, <, bwrap->argv->len);
+  g_assert_cmpstr (bwrap->argv->pdata[i++], ==, NULL);
+  g_assert_cmpuint (i, ==, bwrap->argv->len);
+}
+
+static void
+test_full (void)
+{
+  g_autoptr(FlatpakBwrap) bwrap = flatpak_bwrap_new (NULL);
+  g_autoptr(FlatpakExports) exports = flatpak_exports_new ();
+  gsize i;
+
+  flatpak_exports_add_host_etc_expose (exports,
+                                       FLATPAK_FILESYSTEM_MODE_READ_WRITE);
+  flatpak_exports_add_host_os_expose (exports,
+                                      FLATPAK_FILESYSTEM_MODE_READ_ONLY);
+  flatpak_exports_add_path_expose (exports,
+                                   FLATPAK_FILESYSTEM_MODE_READ_WRITE,
+                                   "/tmp");
+  flatpak_exports_add_path_expose (exports,
+                                   FLATPAK_FILESYSTEM_MODE_READ_ONLY,
+                                   "/var");
+  flatpak_exports_add_path_tmpfs (exports, "/var/tmp");
+  flatpak_exports_add_path_expose_or_hide (exports,
+                                           FLATPAK_FILESYSTEM_MODE_NONE,
+                                           "/home");
+  flatpak_exports_add_path_expose_or_hide (exports,
+                                           FLATPAK_FILESYSTEM_MODE_READ_ONLY,
+                                           "/srv");
+
+  flatpak_bwrap_add_arg (bwrap, "bwrap");
+  flatpak_exports_append_bwrap_args (exports, bwrap);
+  flatpak_bwrap_finish (bwrap);
+  print_bwrap (bwrap);
+
+  i = 0;
+  g_assert_cmpuint (i, <, bwrap->argv->len);
+  g_assert_cmpstr (bwrap->argv->pdata[i++], ==, "bwrap");
+
+  /* Hiding /home just uses --dir because / is not exposed.
+   * We don't create a FAKE_MODE_DIR in the container unless it exists
+   * on the host. (FIXME: is this right?) */
```

Comment:
Gor hiding we create a FAKE_MODE_TMPFS at /home, but during expansion in flatpak_exports_append_bwrap_args() we get to:
https://github.com/flatpak/flatpak/blob/master/common/flatpak-exports.c#L278

Where we avoid mounting a new tmpfs unnecessary if the parent dir is already a tmpfs (due to not being mapped)

Reply:
Sorry for the confusion: this comment was meant to be about `flatpak_exports_add_path_dir()`, but I'm not currently testing that at all.

I agree that for `FAKE_MODE_TMPFS` this behaviour makes sense.

For `FAKE_MODE_DIR`, it seemed a little odd to be skipping `mkdir /path` when `/path` doesn't exist on the host system.

Reply:
I removed the comment for now. Ideally we'd have test coverage for `flatpak_exports_add_path_dir()` too, but I think that can wait until I expand the test to create a temp directory so that we control the nature of the directories we're mounting (exists or not, symlink or not, etc.).